doc/what-is-ppx.rst

```diff
@@ -0,0 +1,119 @@
+************
+What is PPX?
+************
+
+Overview
+--------
+
+Ppx is a meta-programming system for the OCaml programming
+language. It allows developpers to generate code at compile time in a
+principled way. The distinguishing feature of ppx is that it is
+tightly integrated with the OCaml parser and instead of operating at
+the text level it operates on the internal strcutured representation
+of the language in the compiler, called the Abstract Syntax Tree or
+AST for short.
+
+A few years ago, the OCaml language was extended with two new
+constructions: annotations and extension points. Annotations are
+abritrary pieces of information that can be attached to most parts of
+the OCaml language. They can be used to control the behavior of the
+OCaml compiler, or in some specific cases to generate code at compile
+time.
+
+Extension points are compile time functions. The compiler itself
+doesn't know how to interpret them and they must all be rewritten by
+the ppx system before the compiler can process input files further.
+
+Ppxlib mainly supports two ways of generating code at compile time: by
+expanding an extension point or by expanding a ``[@@deriving ...]``
+attribute after a type declaration.
+
+How does it works?
+------------------
+
+The ppx system is composed of 3 parts:
+
+- individual ppx rewriters
+- ppxlib
+- a hook in the compiler
+
+Inidividual ppx rewriters are those implemented by various developpers
+to provide features to end users, such as ppx_expect_ which provides a
+good inline testing framework.
+
+All these rewriters are written against the ppxlib API. Ppxlib is
+responsible for acknowledging the various rewriters a end user wants
+to use, making sure they can be composed together and performing the
+actual rewriting of input files.
+
+The hook in the compiler allows ppxlib to insert itself in the
+compilation pipeline and perform the rewriting of input files based on
+a list of ppx rewriters specified by the user. The hooks take the form
+of command line flags that takes a command to execute. The compiler
+supports two slightly different flags, for providing commands that are
+executed at different stages: ``-pp`` and ``-ppx``. The difference
+between the two is as follow:
+
+- ``-pp`` takes as argument a command that is used to parse the
+  textual representation. Such a command can produce either a plain
+  OCaml source file or a serialised representation of the AST
+
+- ``-ppx`` takes as argument a command that is given a serialised
+  representation of the AST and produces another serialised AST
+
+Ppxlib generally uses the first one as it yields faster compilation
+times, however it supports both methods of operation.
+
+Is ppxlib necessary?
```

Comment:
"necessary" seems to be quite a strong word, given
that the next paragraph states that one can write ppx
rewriters without it.